### PR TITLE
[Example] add entity_sample example of using `CuGraphRelGraphConv`

### PR DESCRIPTION
**Pull request by tingyu66**
## Description

This PR adds an example of using `dgl.nn.CuGraphRelGraphConv`.

Needs #4826 to be merged first. Partially address #4669.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            logits = model(blocks)
+            eval_logits.append(logits.cpu().detach())
+            eval_seeds.append(output_nodes.cpu().detach())
+    eval_logits = torch.cat(eval_logits)
+    eval_seeds = torch.cat(eval_seeds)
+    return accuracy(eval_logits.argmax(dim=1), labels[eval_seeds].cpu()).item()
+
+
+def train(device, g, target_idx, labels, train_mask, model, fanouts):
+    # define train idx, loss function and optimizer
```
Comment (frozenbugs) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:65]: nit: for all comment. Start with capital and end with dot.

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    else:
+        raise ValueError("Unknown dataset: {}".format(args.dataset))
+    hg = data[0].to(device)
+    hg = hg.int() if args.idtype == "int32" else hg.long()
+    num_rels = len(hg.canonical_etypes)
+    category = data.predict_category
+    labels = hg.nodes[category].data.pop("labels")
+    train_mask = hg.nodes[category].data.pop("train_mask")
+    test_mask = hg.nodes[category].data.pop("test_mask")
+    # find target category and node id
```
Comment (frozenbugs) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:147]: nit: Add a blank line to give the code some space to breathe?

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    category_id = hg.ntypes.index(category)
+    g = dgl.to_homogeneous(hg)
+    node_ids = torch.arange(g.num_nodes()).to(device)
+    target_idx = node_ids[g.ndata[dgl.NTYPE] == category_id]
+    # rename the field as they can be changed by DataLoader
```
Comment (frozenbugs) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:152]: Can you be more specific?

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    # create RGCN model
+    in_size = g.num_nodes()  # featureless with one-hot encoding
+    out_size = data.num_classes
+    num_bases = 20
+    fanouts = [4, 4]
+    model = RGCN(in_size, 16, out_size, num_rels, num_bases, fanouts).to(device)
+
+    train(device, g, target_idx, labels, train_mask, model, fanouts)
+    test_idx = torch.nonzero(test_mask, as_tuple=False).squeeze()
+    # Note: cugraph-ops aggregators are designed for sampled graphs (MFGs) and expect fanout
```
Comment (frozenbugs) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:170]: over 80 char.

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py (hunk 3 above)
Comment (tingyu66) [reply examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py]: This was copied from entity_sample directly. Comment has been removed.

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by mufeili** [comment]
How about placing this example with `examples/cugraph/rgcn.py`? @tingyu66 @frozenbugs @jermainewang

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="RGCN for entity classification with sampling"
+    )
+    parser.add_argument(
+        "--dataset",
+        type=str,
+        default="aifb",
+        help="Dataset name ('aifb', 'mutag', 'bgs', 'am'), default to 'aifb'.",
```
Comment (mufeili) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:118]:
```suggestion
        choices=['aifb', 'mutag', 'bgs', 'am'],
```

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    )
+    parser.add_argument(
+        "--idtype",
+        type=str,
+        default="int64",
+        help="Graph idtype ('int64' or 'int32'), default to 'int64'.",
+    )
```
Comment (mufeili) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:125]: Can we remove this argument and stick to one particular option? For users, this seems to be irrelevant for using cuGraph.

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    if args.dataset == "aifb":
+        data = AIFBDataset()
+    elif args.dataset == "mutag":
+        data = MUTAGDataset()
+    elif args.dataset == "bgs":
+        data = BGSDataset()
+    elif args.dataset == "am":
+        data = AMDataset()
+    else:
+        raise ValueError("Unknown dataset: {}".format(args.dataset))
```
Comment (mufeili) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:140]:
```suggestion
        raise ValueError(f"Unknown dataset: {args.dataset}")
```

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    test_loader = DataLoader(
+        g,
+        target_idx[test_idx].type(g.idtype),
+        test_sampler,
+        device=device,
+        batch_size=32,
+        shuffle=False,
+    )
+    acc = evaluate(model, labels, test_loader, inv_target)
+    print("Test accuracy {:.4f}".format(acc))
```
Comment (mufeili) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:187]:
```suggestion
    print(f"Test accuracy {acc:.4f}")
```

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    # No separate validation subset, use train index instead for validation.
+    val_loader = DataLoader(
+        g,
+        target_idx[train_idx].type(g.idtype),
+        sampler,
+        device=device,
+        batch_size=100,
+        shuffle=False,
+    )
+    model.train()
```
Comment (mufeili) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:89]: Move this to L91 since you call `model.eval()` in `evaluate`

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+            logits = model(blocks)
+            loss = loss_fcn(logits, labels[output_nodes])
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            total_loss += loss.item()
+        acc = evaluate(model, labels, val_loader, inv_target)
+        print(
+            "Epoch {:05d} | Loss {:.4f} | Val. Accuracy {:.4f} ".format(
+                epoch, total_loss / (it + 1), acc
```
Comment (mufeili) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:105]: Use f-strings.

**Review by mufeili — CHANGES_REQUESTED** [review]
Some minor comments

**Comment by tingyu66** [comment]
> How about placing this example with `examples/cugraph/rgcn.py`? @tingyu66 @frozenbugs @jermainewang

I have no objection. The current name is indeed a bit lengthy.

**Comment by frozenbugs** [comment]
As we discussed in https://docs.google.com/document/d/1oVBrYPQFUYnel0zK69yTiRWvoRvLeOpCWHazynT9bpM/edit#, 
If we have multiple cugraph examples, then: examples/advanced/cugraph/rgcn.py
Otherwise: examples/advanced/rgcn_cugraph.py

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
+import torch
```
Comment (frozenbugs) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:1]: Add a concise description at the top of the file?
https://github.com/dmlc/dgl/blob/master/examples/sparse/sign.py

Comment (frozenbugs) [reply examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:1]: Highlight the "what's new" and "key part that you want to sell the users" in this example in the description or in the code.

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from torchmetrics.functional import accuracy
```
Comment (frozenbugs) [inline examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:4]: Do we have to use torchmetrics?
Is torchmetrics a dependency of DGL?
@mufeili

Comment (mufeili) [reply examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py:4]: It's not a dependency. However @jermainewang previously suggested using it so we can avoid writing some lines of code for metric computation ourselves.

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py (hunk 11 above)
Comment (tingyu66) [reply examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py]: Description added and the example now goes into the advanced directory.

File: examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py (hunk 12 above)
Comment (jermainewang) [reply examples/pytorch/rgcn/entity_sample_cugraph_relgraphconv.py]: Yeah, it saves many codes for computing standard metrics like accuracy.

**Review by mufeili — APPROVED** [review]

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot
